ansible/roles-infra/infra-ec2-template-generate/tasks/validate_instances_amis.yaml

```diff
@@ -0,0 +1,116 @@
+  - name: test 1
+    debug: msg="{{ instances }}"
+
+  - name: prep facts
+    set_fact:
+      instances_DoD: "{{ [] }}"  # DoD
+  
+  - name: fail if I can't map an image to aws_ami_region_mapping
+    loop: "{{ instances }}"
+    when:
+    - item.image is defined
+    - aws_ami_region_mapping[aws_region][item.image] is not defined
+    fail:
+      msg: >- 
+        instance.image "{{ item.image }}" is not defined in aws_ami_region_mapping.
+
+  - name: create instances_DoD from instances LoD
+    loop: "{{ instances }}"
+    loop_control:
+      extended: yes
+    set_fact: 
+      instances_DoD: >-
+        {{ 
+          instances_DoD| combine(
+            { ansible_loop.index0: instances[ansible_loop.index0]  }
+          )
+        }}
+
+  - name: test if bump RHEL<VERSION> images to GOLD, if available in aws_ami_region_mapping
+    loop: "{{ instances_DoD|dict2items }}"
+    when: 
+      - item.value['image'] is defined
+      - item.value['image'] is regex("^RHEL.*$")
+      - item.value['image'] is not regex("^.*GOLD$")      
+      - aws_ami_region_mapping[aws_region][item.value['image']|regex_replace("^(.*)$", "\\1GOLD")] is defined
+    ec2_ami_facts:
+      # profile: "{{ profile }}"
+      region: "{{ aws_region }}"
+      image_ids: >-
+        {{ aws_ami_region_mapping[aws_region][
+            item.value['image']|regex_replace("^(.*)$", "\1GOLD")
+          ] }}
+    register: ami_presence
+
+  - name: debug ami_presence
+    debug:
+      msg: "{{ ami_presence.results }}"
+
+  - name: rename successful bumps from RHEL<VERSION> to RHEL<VERSION>GOLD in instances_DoD
+    # items that returned images array not empty
+    loop: "{{ ami_presence.results|json_query(' [? not_null(invocation) && not_null(images)].item.key' ) }}"
+    set_fact:
+      instances_DoD: >-
+        {{ instances_DoD| combine(
+          { item:
+            { "image": instances_DoD[item]['image']|regex_replace("^(.*)$", "\1GOLD")}
+            }, 
+          recursive=True
+        )}}
```

Comment:
So the idea is to  update a key in a dict that is part of a list. (`instances`).

Why don't we come up with a new dedicated variable instead ?

```dynamic_instances_image_promotion:
  instance_name:  RHEL77GOLD
```
And then in the template, instead of doing:

```yaml
      ImageId:
        Fn::FindInMap:
        - RegionMapping
        - Ref: AWS::Region
        - {{ instance.image | default(aws_default_image) }}
```

We do

```yaml
      ImageId:
        Fn::FindInMap:
        - RegionMapping
        - Ref: AWS::Region
        - {{ dynamic_instances_image_promotion[instance.name] | default(instance.image) | default(aws_default_image) }}
```

That would require to update all the templates.
But we will have to update all of them since in this code we're using aws_ami_region_mapping  mapping and it's not always included..

```
lolo:{fix-test-bool-when} ~/sync/dev/aad/ansible$ find -name 'ec2_cloud_template.*'| while read i; do if ! grep -q aws_ami_region_mapping $i; then echo $i;fi; done
./configs/ocp-ha-lab/files/cloud_providers/ec2_cloud_template.j2
./configs/ansible-provisioner/files/cloud_providers/ec2_cloud_template.j2
./configs/three-tier-app/files/cloud_providers/ec2_cloud_template.j2
./configs/ansible-cicd-lab/files/cloud_providers/ec2_cloud_template.j2
./configs/ocp-ha-disconnected-lab/files/cloud_providers/ec2_cloud_template.j2
./configs/ocp-gpu-single-node/files/cloud_providers/ec2_cloud_template.j2
./configs/ansible-tower-implementation/files/cloud_providers/ec2_cloud_template.j2
./configs/rhte-oc-cluster-vms/files/cloud_providers/ec2_cloud_template.j2
./configs/ocp-implementation-lab/files/cloud_providers/ec2_cloud_template.j2
./configs/ocp-storage-cns/files/cloud_providers/ec2_cloud_template.j2
./configs/ansible-windows/files/cloud_providers/ec2_cloud_template.j2
./configs/ans-tower-ops-demo/files/cloud_providers/ec2_cloud_template.j2
./configs/ans-network-lab/files/cloud_providers/ec2_cloud_template.j2
./configs/amq-messaging-foundations/files/cloud_providers/ec2_cloud_template.j2
./configs/ocp-workshop-with-clientvms/files/cloud_providers/ec2_cloud_template.j2
./configs/ans-tower-lab/files/cloud_providers/ec2_cloud_template.j2
./configs/ansible-integrations/files/cloud_providers/ec2_cloud_template.j2
./configs/test-empty-config/files/cloud_providers/ec2_cloud_template.j2
```

I expect all those configs to fail with your code, since the mapping doesn't exist in the templates ?

Reply:
It does work, because all code has access to the `aws_ami_region_mapping` even if not used in the template.

If the template doesn't process instance.image (which many still don't, as above) then nothing is changed.

Template that do not use them, ignore them.